ksql-engine/src/test/java/io/confluent/ksql/planner/plan/KsqlStructuredDataOutputNodeTest.java

```diff
@@ -255,14 +281,105 @@ public void shouldCreateSinkWithCorrectCleanupPolicyWindowedTable() {
+  }
+
+  @Test
+  public void shouldCreateSinkWithTheSourcePartitionReplication() {
+    // Given:
+    createOutputNode(Collections.emptyMap(), true);
+
+    // When:
+    final SchemaKStream schemaKStream = buildStream();
+
+    // Then:
+    verify(mockTopicClient).createTopic(SINK_KAFKA_TOPIC_NAME, 1, (short) 2, Collections.emptyMap());
+    assertThat(schemaKStream, instanceOf(SchemaKStream.class));
+  }
+
+  @Test
+  public void shouldNotFetchSourceTopicPropsIfProvided() {
+    // Given:
+    createOutputNode(ImmutableMap.of(
+        KsqlConfig.SINK_NUMBER_OF_PARTITIONS_PROPERTY, 5,
+        KsqlConfig.SINK_NUMBER_OF_REPLICAS_PROPERTY, (short) 3
+    ), true);
+
+    // When:
+    final SchemaKStream schemaKStream = buildStream();
+
+    // Then:
+    verify(mockTopicClient, never()).describeTopics(any());
+    verify(mockTopicClient).createTopic(SINK_KAFKA_TOPIC_NAME, 5, (short) 3, Collections.emptyMap());
+    assertThat(schemaKStream, instanceOf(SchemaKStream.class));
+  }
+
+  @Test
+  public void shouldCreateSinkWithTheSourceReplicationAndProvidedPartition() {
+    // Given:
+    createOutputNode(Collections.singletonMap(KsqlConfig.SINK_NUMBER_OF_PARTITIONS_PROPERTY, 5), true);
+
+    // When:
+    final SchemaKStream schemaKStream = buildStream();
+
+    // Then:
+    verify(mockTopicClient).createTopic(SINK_KAFKA_TOPIC_NAME, 5, (short) 2, Collections.emptyMap());
+    assertThat(schemaKStream, instanceOf(SchemaKStream.class));
+  }
+
+  @Test
+  public void shouldCreateSinkWithTheSourcePartitionAndProvidedReplication() {
+    // Given:
+    createOutputNode(Collections.singletonMap(KsqlConfig.SINK_NUMBER_OF_REPLICAS_PROPERTY, (short) 2), true);
+
+    // When:
+    final SchemaKStream schemaKStream = buildStream();
+
+    // Then:
+    verify(mockTopicClient).createTopic(SINK_KAFKA_TOPIC_NAME, 1, (short) 2, Collections.emptyMap());
+    assertThat(schemaKStream, instanceOf(SchemaKStream.class));
+  }
+
+  @Test
+  public void shouldThrowIfSinkTopicHasDifferentPropertiesThanRequested() {
+    // Given:
+    doThrow(KsqlException.class).when(mockTopicClient).createTopic(SINK_KAFKA_TOPIC_NAME, 1, (short) 2, Collections.emptyMap());
+    createOutputNode(Collections.singletonMap(KsqlConfig.SINK_NUMBER_OF_REPLICAS_PROPERTY, (short) 2), true);
+    expectedException.expect(KsqlException.class);
+
+    // When:
+    buildStream();
+
+  }
+
+  @Test
+  public void shouldUseLegacySinkPropertiesIfLegacyIsTrue() {
+    // Given:
+    Mockito.<Object>when(ksqlConfig.values()).thenReturn((Map<String, ?>) ImmutableMap.of(
+        KsqlConfig.SINK_NUMBER_OF_PARTITIONS_PROPERTY, KsqlConstants.defaultSinkNumberOfPartitions,
+        KsqlConfig.SINK_NUMBER_OF_REPLICAS_PROPERTY, KsqlConstants.defaultSinkNumberOfReplications
+    ));
+    when(ksqlConfig.getInt(KsqlConfig.SINK_NUMBER_OF_PARTITIONS_PROPERTY)).thenReturn(KsqlConstants.defaultSinkNumberOfPartitions);
```

Comment:
weird double mocking of `KsqlConfig.SINK_NUMBER_OF_PARTITIONS_PROPERTY` in this test - but I think this will go away once you simplify the if statement in the prod code this is testing.